File: next/Journal/FactCodec.fs
namespace Wanxiangshu.Next.Journal

open System
open Thoth.Json
open Wanxiangshu.Next.Kernel.Fact

/// Fact serialization (PERSIST-005).
module FactCodec =

    let private extra = Extra.empty |> Extra.withInt64

    let pre050MigrationMessage =
        "Wanxiangshu 0.5.0 does not support pre-0.5.0 runtime journals.\nArchive or remove the old Wanxiangshu runtime journal before starting."

    /// Field and case names that only a pre-0.5.0 journal can contain.
    ///
    /// Two groups, both fatal:
    ///
    /// Fields — the old Fallback projection stored Dead/Failures counters and
    /// model ids. Guessing a modulo-4 cursor from them would be inventing
    /// history, and VERIFY-006 lists a journal carrying model ids as a No-Go.
    ///
    /// Case names — the facts replaced in this migration. Without them the
    /// decoder would fail with an opaque union error, so the operator would see
    /// "cannot parse line 3" instead of "this journal predates 0.5.0". A precise
    /// diagnosis is the difference between archiving the file and debugging the
    /// codec.
    let private pre050Markers =
        [| "\"FailuresOnCurrentSide\""
           "\"IsDead\""
           "\"TotalFailures\""
           "\"BaseModelID\""
           "\"BaseProviderID\""
           "\"EffectiveModelID\""
           "\"EffectiveProviderID\""
           "\"PluginPromptAccepted\""
           "\"HumanPromptAccepted\""
           "\"GuardPromptAccepted\""
           "\"InteractionRepairClaimed\""
           "\"ReviewConfirmedIdle\""
           "\"AgentLinked\""
           "\"AgentForked\""
           "\"AgentUnlinked\""
           "\"OrchestratorManagerJobCreated\""
           "\"OrchestratorCandidateRegistered\""
           "\"OrchestratorPublished\""
           "\"OrchestratorRejected\""
           "\"OrchestratorRebased\""
           "\"OrchestratorConflictDetected\""
           "\"OrchestratorPreRebaseReviewConfirmed\""
           "\"OrchestratorPostRebaseReviewConfirmed\""
           "\"OrchestratorPublishClaimed\"" |]

    let containsLegacyFallbackFields (json: string) =
        pre050Markers
        |> Array.exists (fun marker -> json.IndexOf(marker, StringComparison.Ordinal) >= 0)

    let serializeFact (fact: Fact) : string =
        Encode.Auto.toString (0, fact, extra = extra)

    let deserializeFact (json: string) : Result<Fact, string> =
        if containsLegacyFallbackFields json then
            Error pre050MigrationMessage
        else
            Decode.Auto.fromString<Fact> (json, extra = extra)

File: .wanxiangshu/runtimes/Journal/FactCodec.js

import { newGuid } from "../fable_modules/fable-library-js.5.6.0/Guid.js";
import { add } from "../fable_modules/fable-library-js.5.6.0/Map.js";
import { Auto_generateBoxedEncoder_437914C6, toString, int64 } from "../fable_modules/Thoth.Json.10.5.1/Encode.fs.js";
import { Auto_generateBoxedDecoder_Z6670B51, fromString, int64 as int64_1 } from "../fable_modules/Thoth.Json.10.5.1/Decode.fs.js";
import { empty } from "../fable_modules/Thoth.Json.10.5.1/Extra.fs.js";
import { ExtraCoders } from "../fable_modules/Thoth.Json.10.5.1/Types.fs.js";
import { Fact_$reflection } from "../Kernel/Fact.js";
import { uncurry2 } from "../fable_modules/fable-library-js.5.6.0/Util.js";
import { FSharpResult$2 } from "../fable_modules/fable-library-js.5.6.0/Result.js";

const extra = new ExtraCoders((() => {
    let copyOfStruct = newGuid();
    return copyOfStruct;
})(), add("System.Int64", [int64, int64_1], empty.Coders));

export function serializeFact(fact) {
    return toString(0, Auto_generateBoxedEncoder_437914C6(Fact_$reflection(), undefined, extra, undefined)(fact));
}

export function deserializeFact(json) {
    const matchValue = fromString(uncurry2(Auto_generateBoxedDecoder_Z6670B51(Fact_$reflection(), undefined, extra)), json);
    if (matchValue.tag === 1) {
        return new FSharpResult$2(1, [matchValue.fields[0]]);
    }
    else {
        return new FSharpResult$2(0, [matchValue.fields[0]]);
    }
}

